### DISCUSSION TITLE
HugeGraph Integrates with LangChain(LLM application framework)

### DISCUSSION THREAD

**Discussion by simon824** [apache/incubator-hugegraph#2251] [Ideas]
[LangChain](https://github.com/hwchase17/langchain) is a framework for developing applications powered by language models. It enables applications that are:
- Data-aware: connect a language model to other sources of data
- Agentic: allow a language model to interact with its environment

We created https://github.com/hwchase17/langchain/pull/7132 to make HugeGraph integrates with LangChain

We can：
1. Use Langchain to combine HugeGraph's schema and natural language to generate gremlin/cypher , and then execute query operations on HugeGraph.
2. Use the result returned by hugegraph to supplement the context of the answer, and send it to LLM as a prompt.

`User document` about using hugegraph in  langchain :
https://python.langchain.com/docs/modules/chains/additional/graph_hugegraph_qa

Let's discuss here if you are interested in the integration of graph databases and large language models.

**Comment by liuxiaocs7** [comment]
Successfully reproduced, this feature will be very helpful for writing Gremlin statements.

![image](https://github.com/apache/incubator-hugegraph/assets/42756849/977c3199-69b9-4c81-8bb1-67bc3581c653)

Some points to note:

1. some dependencies that need to be installed:

```
requests
decorator
```

2. If we want to use openai, we also need to install

```
openai
```

and set our own api-key in `OPENAI_API_KEY`.

---

HG Py-Client: https://pypi.org/project/hugegraph-python/, looking forward to update.

- [Langchain Beginners Getting Started Guide | LangChain初学者入门指南](https://mp.weixin.qq.com/s/F4QokLPrimFS1LRjXDbwQQ)
